src/Microsoft.Health.Fhir.Shared.Api/Controllers/DataConvertController.cs

```diff
@@ -0,0 +1,179 @@
+            if (referenceComponents.Length <= 1 || string.IsNullOrWhiteSpace(referenceComponents.First()))
+            {
+                _logger.LogInformation("Templates collection reference is invalid: registry server missing.");
+                throw new RequestNotValidException(string.Format(Resources.InvalidTemplateCollectionReference, templateCollectionReference));
+            }
+
+            return referenceComponents[0];
+        }
+
+        private void ValidateParams(Parameters inputParams)
+        {
+            if (inputParams == null)
+            {
+                _logger.LogInformation("Failed to deserialize data convert request body as Parameters resource.");
+                throw new RequestNotValidException(Resources.DataConvertParametersNotValid);
+            }
+
+            var supportedParams = _supportedParams[Request.Method];
+
+            foreach (var param in inputParams.Parameter)
+            {
+                var paramName = param.Name;
+                if (!supportedParams.Contains(paramName))
+                {
+                    throw new RequestNotValidException(string.Format(Resources.DataConvertParameterNotValid, paramName));
+                }
+            }
+        }
+
+        private static string ReadStringParameter(Parameters parameters, string paramName)
+        {
+            var param = parameters?.Parameter.Find(p =>
+                string.Equals(p.Name, paramName, StringComparison.OrdinalIgnoreCase));
+
+            var paramValue = param?.Value?.ToString();
+            if (string.IsNullOrEmpty(paramValue))
+            {
+                throw new RequestNotValidException(string.Format(Resources.DataConvertParameterValueNotValid, paramName));
+            }
+
+            return paramValue;
+        }
+
+        private static T ReadEnumParameter<T>(Parameters parameters, string paramName)
+        {
+            var param = parameters?.Parameter.Find(p =>
+                string.Equals(p.Name, paramName, StringComparison.OrdinalIgnoreCase));
+
+            object enumValue;
+            if (!Enum.TryParse(typeof(T), param?.Value?.ToString(), ignoreCase: true, out enumValue))
+            {
+                throw new RequestNotValidException(string.Format(Resources.DataConvertParameterValueNotValid, paramName));
+            }
+
+            return (T)enumValue;
+        }
+
+        private static Dictionary<string, HashSet<string>> InitSupportedParams()
+        {
+            var postParams = new HashSet<string>()
+            {
+                DataConvertProperties.InputData,
+                DataConvertProperties.InputDataType,
+                DataConvertProperties.TemplateCollectionReference,
+                DataConvertProperties.EntryPointTemplate,
+            };
+
+            var supportedParams = new Dictionary<string, HashSet<string>>();
+            supportedParams.Add(HttpMethods.Post, postParams);
```

Comment:
>supportedParams.Add(HttpMethods.Post, postParams) [](start = 12, length = 49)

Is there are plans to add more actions for this controller?
Right now it's looks a bit odd to have dictionary with only one key, naturally I question: Can we avoid dictionary in this case and just use hashset?

Reply:
Currently no other actions planned. I have simplified the structure to HashSet.